pallene/coder.lua

```diff
@@ -1437,6 +1437,42 @@ gen_cmd["BuiltinIoWrite"] = function(self, cmd, _func)
     return util.render([[ pallene_io_write(L, $v); ]], { v = v })
 end
 
+gen_cmd["BuiltinMathAbs"] = function(self, cmd, _func)
+    local dst = self:c_var(cmd.dsts[1])
+    local v = self:c_value(cmd.srcs[1])
+    return util.render([[ $dst = l_mathop(fabs)($v); ]], { dst = dst, v = v })
+end
+
+gen_cmd["BuiltinMathExp"] = function(self, cmd, _func)
+    local dst = self:c_var(cmd.dsts[1])
+    local v = self:c_value(cmd.srcs[1])
+    return util.render([[ $dst = l_mathop(exp)($v); ]], { dst = dst, v = v })
+end
+
+gen_cmd["BuiltinMathLog"] = function(self, cmd, _func)
+    local dst = self:c_var(cmd.dsts[1])
+    local v = self:c_value(cmd.srcs[1])
+    local b = self:c_value(cmd.srcs[2])
+	local base_num = tonumber(b)
+    if base_num == 1.0 then
```

Comment:
Technically, using `1.0` to stand for `e` would be a break in compatibility. As a rule of thumb we try to ensure that when a Pallene program runs without errors, it produces the same result that Lua would.

However, it might be acceptable as a temporary workaround, until we implement proper optional arguments. If we choose to go forward with this workaround, I would recommend adding a code comment explaining it, and creating a separate issue in the issue tracker, so we do not forget to fix it later.

Reply:
I think 1.0 is what Lua does. Below is the code from lmathlib.c Notably, the first case uses lua_isoneornil(L, 2) to trigger the log() case. So, by my understanding, if the parameter is either 1.0 (explicit parameter) or nil (omitted optional parameter), log() is used.

static int math_log (lua_State *L) {
  lua_Number x = luaL_checknumber(L, 1);
  lua_Number res;
  if (lua_isnoneornil(L, 2))
    res = l_mathop(log)(x);
  else {
    lua_Number base = luaL_checknumber(L, 2);
#if !defined(LUA_USE_C89)
    if (base == l_mathop(2.0))
      res = l_mathop(log2)(x);
    else
#endif
    if (base == l_mathop(10.0))
      res = l_mathop(log10)(x);
    else
      res = l_mathop(log)(x)/l_mathop(log)(base);
  }
  lua_pushnumber(L, res);
  return 1;
}

Reply:
It's a subtle difference of only one character... Lua is using "none or nil", not "one or nil". It checks if the second argument is absent, or if it is nil. That is, `math.log(x)` or `math.log(x, nil)`. While in regular Lua these two are alsmost indistinguishable, inside the C API  there is a distinction between missing argument and nil argument.

```
> math.log(3.0, 1.0)
inf
```